### PR TITLE
[run_process_replay] linearizer: fix get_grouping_dims to respect global/local max 

### PR DESCRIPTION
**Pull request by flammit**

### CODE REVIEW

File: tinygrad/renderer/cstyle.py [hunk 1]
```diff
@@ -220,6 +220,7 @@ def __init__(self): self.tensor_cores = MetalRenderer.tensor_cores if os.uname()
   float4 = "float4"
   uses_ptr_arithmetic = True
   code_for_workitem = {"g": lambda x: f"gid.{chr(120+x)}", "l": lambda x: f"lid.{chr(120+x)}"}
+  # uint3 used for gid/lid - TODO: this should probably be `ushort3 lid [[thread_position_in_threadgroup]]`
```
Comment (Qazalin) [inline tinygrad/renderer/cstyle.py:223]: yeah I remember RDNA3 reused VGPRs with bitshifting for lidx but not for gidx - maybe the compilers just handle it

File: tinygrad/codegen/linearizer.py [hunk 2]
```diff
@@ -379,25 +420,22 @@ def linearize(self) -> Linearizer:
...
+    # render global and local as specials or a loop
+    if self.opts.has_local:
+      self.loop_uops.update({x.expr:UOp(UOps.SPECIAL, dtypes.int32, (), (i, x.expr, x.max+1)) for i,x in enumerate(loop_global_idxs)})
+      if not self.dont_use_locals:
+        self.loop_uops.update({x.expr:UOp(UOps.SPECIAL, dtypes.int32, (), (i, x.expr, x.max+1)) for i,x in enumerate(loop_local_idxs)})
     else:
+      self.global_size, self.local_size = None, None
       self.render_loop(loop_global_idxs+loop_local_idxs, 1)
-    if self.global_size is not None: self.global_size += [1]*(3-len(self.global_size))
-    if self.local_size is not None: self.local_size += [1]*(3-len(self.local_size))
```
Comment (Qazalin) [inline tinygrad/codegen/linearizer.py:400]: can self.global_size and local_size be tuples now? we're not changing them anymore

**Comment by Qazalin** [comment]
Not sure if this was intentional but looks like it also cleverly reorders indexing ALUs based on scope!

https://github.com/tinygrad/tinygrad/actions/runs/9521487566/job/26249009174#step:23:55

But the process replay output mixes up these changes with lidx starting at 0

https://github.com/tinygrad/tinygrad/actions/runs/9521608695/job/26249418646#step:23:50

can you keep it as before to separate signal from noise in the process replay diff?

File: tinygrad/codegen/linearizer.py (hunk 2 above)
Comment (flammit) [reply tinygrad/codegen/linearizer.py:400]: yup will fix that!

Comment (flammit) [reply tinygrad/codegen/linearizer.py:400]: actually changing to tuple is a lot more difficult as it's used in a lot of places and search actually modifies the sizes.

**Comment by Qazalin** [comment]
cool, thanks @flammit!
would you wanna get this merged or delete the old way here too?

**Comment by flammit** [comment]
> cool, thanks @flammit! would you wanna get this merged or delete the old way here too?

probably better to merge this as is, then I'll push another PR that changes the behavior and we can verify the benchmarks on that?

gimme a bit though - I'm gonna retest unet3d so make sure @francislata's issue is resolved.

**Comment by flammit** [comment]
@Qazalin looks like unet3d worked fine.  rebased to master and good to merge if you don't see any other issues.

I'll post a separate, follow-on PR tomorrow to do the behavioral changes that are not replay safe (direction of variable indexing and which side of the sizes (.x or .z) to accumulate extra dims into first.

File: tinygrad/codegen/linearizer.py (hunk 2 above)
Comment (Qazalin) [reply tinygrad/codegen/linearizer.py:400]: oh ok cool

**Pull request merged**